lib/active_shipping/carriers/usps.rb

```diff
@@ -336,6 +408,171 @@ def canned_address_verification_works?
       xml.at('/CarrierPickupAvailabilityResponse/City').try(:text) == 'SAN FRANCISCO' && xml.at('/CarrierPickupAvailabilityResponse/Address2').try(:text) == '18 FAIR AVE'
     end
 
+    def build_shipment_request(action, origin, destination, package, options = {})
+      if us_address?(destination)
+        build_us_shipment_request(action, origin, destination, package, options)
+      else
+        build_world_shipment_request(action, origin, destination, package, options)
+      end
+    end
+
+    def build_us_shipment_request(action, origin, destination, package, options = {})
+      xml_builder = Nokogiri::XML::Builder.new do |xml|
+        xml.send(XML_ROOTS[action].to_sym, 'USERID' => @options[:login]) do
+          service = SHIPMENT_SERVICES[options[:service]]
+
+          unless service
+            raise ArgumentError, "Service is not provided or not supported."
+          end
+
+          build_us_location_node(xml, 'From', origin)
+          build_us_location_node(xml, 'To', destination)
+
+          xml.WeightInOunces("%0.1f" % [package.ounces, 1].max)
+          xml.ServiceType(service)
+
+          xml.ImageType(LABEL_TYPE)
+
+          size_code = USPS.size_code_for(package)
+          container = CONTAINERS[package.options[:container]]
+          container ||= (package.cylinder? ? 'NONRECTANGULAR' : 'RECTANGULAR') if size_code == 'LARGE'
+          xml.Container(container)
+
+          xml.Size(size_code)
+        end
+      end
+
+      save_request(xml_builder.to_xml)
+    end
+
+    def build_us_location_node(xml, prefix, address)
+      array = [
+        ['Name', address.name],
+        ['Firm', address.company],
+        ['Address1', address.address1],
+        ['Address2', address.address2 || address.address1],
+        ['City', address.city],
+        ['State', address.state],
+        ['Zip5', strip_zip(address.zip)]
+      ]
+      array << ['Zip4', strip_zip4(address.zip)] if strip_zip4(address.zip)
+
+      array.each do |key, value|
+        xml.public_send(prefix + key, value)
+      end
+    end
+
+    def build_world_shipment_request(action, origin, destination, package, options = {})
+      xml_builder = Nokogiri::XML::Builder.new do |xml|
+        xml.send(XML_ROOTS[action].to_sym, 'USERID' => @options[:login]) do
+          xml.Revision(WORLD_SHIPMENT_API_REVISION)
+          service = SHIPMENT_SERVICES[options[:service]]
+
+          unless service
+            raise ArgumentError, "Service is not provided or not supported."
+          end
+
+          build_world_location_node(xml, 'From', origin)
+          build_world_location_node(xml, 'To', destination)
+
+          size_code = USPS.size_code_for(package)
+          container = CONTAINERS[package.options[:container]]
+          container ||= (package.cylinder? ? 'NONRECTANGULAR' : 'RECTANGULAR') if size_code == 'LARGE'
+          xml.Container(container)
+
+          xml.ShippingContents do |xml|
```

Comment:
Yeah, we presently have the [class `PackageItem` which is perfect for a customs item within a package.](https://github.com/Shopify/active_shipping/blob/master/lib/active_shipping/package.rb#L5); however, it's not used in `active_shipping` explicitly. 

If you'd like to introduce a `PackageItem` array attribute on `Package`, I'd welcome that - alternatively, you could store it in the options during initialization, e.g. 

```
package_items = [pkg_item1, pkg_item2]
package = Package.new(10, [10, 20, 30], package_items: package_items)
```

This pattern is a little undesirable because it places something relatively important into an `options` Hash but it would work right out of the box.